### PR TITLE
Fix unsynchronized buffer access.

### PR DESCRIPTION
**Pull request by Melraidin**
@karmakaze https://app.asana.com/0/191296555608986/211487888617068

### CODE REVIEW

**Comment by Melraidin** [comment]
@junjizhi

File: lib/px/service/kinesis/base_request.rb [hunk 1]
```diff
@@ -36,6 +36,7 @@ def initialize
       @last_throughput_exceeded = nil
 
       @buffer = []
+      @semaphore = Mutex.new
```
Comment (junjizhi) [inline lib/px/service/kinesis/base_request.rb:39]: Should we make it a class level variable? It doesn't protect the critical region if there are multiple client instances.

File: lib/px/service/kinesis/base_request.rb [hunk 2]
```diff
@@ -89,15 +90,17 @@ def flush_records
     #
     # Returns the number of unsent messages
     def queue_record(data)
-      data_blob = data.to_msgpack
+      @semaphore.synchronize do
```
Comment (junjizhi) [inline lib/px/service/kinesis/base_request.rb:93]: Are we sure it is this race condition causing the duplicate activities?

**Review by junjizhi — CHANGES_REQUESTED** [review]

**Review by junjizhi — APPROVED** [review]

**Comment by junjizhi** [comment]
👍

**Pull request merged**